scripts/build.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/env python
+
+# Copyright 2020 Google
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Assuming pre-requisites are in place (please run `scripts/install_preqreqs.py` once on the system), 
+this script will build the firebase cpp sdk
+
+It does the following,
+- install platform specific cpp dependencies via vcpkg
+- cmake configure
+- cmake build
+
+Usage examples:
+# Build all targets with x64 architecture, Release config and 8 cpu cores fo
+python build.py --arch x64 --config Release -j 8
+
+# Build all targets with default options and also build unit tests
+python build.py --build_tests --arch x64
+
+# Build only firebase_app and firebase_auth with default options and also build unit tests
+python build.py --build_tests --arch x64 --targets firebase_app firebase_auth
+
+"""
+
+import subprocess
+import os
+import sys
+import argparse
+import platform
+
+
+def run_command(cmd, cwd=None, as_root_user=False, collect_output=False):
+ """Run a command.
+
+ Args:
+  cmd (list(str)): Command to run as a list object.
+          Eg: ['ls', '-l'].
+  cwd (str): Directory to execute the command from.
+  as_root_user (bool): Run command as root user with admin priveleges (supported on mac and linux).
+  collect_output (bool): Get output from the command. The string returned from this function
+       call will have the output from command execution.
+ Raises:
+  (subprocess.CalledProcessError): If command errored out.
```

Comment:
There should be an option to ignore errors, possibly useful in prereqs. (though if we move this function to a shared module, we can add functionality to it as we need it).

Reply:
If we are maintaining parity with a shell script, we probably shouldn't ignore the error and stop when a command fails. Also, ignoring on the consumer end could be easily done with a try-except block .

Reply:
Generally we run shell scripts with -e so they will exit on any error. We can tweak this later though.

Reply:
cool. Python 3.7 has a convenient way to deal with this (check=True argument to subprocess.run will raise an exception, else it wont). Apologies but since I restructured the code, it will show up as a brand new file.